plugins/portmap/README.md

```diff
@@ -0,0 +1,103 @@
+## Port-mapping plugin
+
+This plugin will forward traffic from one or more ports on the host to the
+container. It expects to be run as a chained plugin.
+
+## Usage
+You should use this plugin as part of a network configuration list. It accepts
+the following configuration options:
+
+* `noSnat` - boolean, defaul false. If true, do not set up the SNAT chains
+
+The plugin expects to receive the actual list of port mappings via the 
+[`portMappings` capability argument](CONVENTIONS.md#plugin-specific-fields)
+
+So a typical config may look like,
+
+```json
+{
+        "cniVersion": "0.3.0",
+        "name": "mynet",
+        "plugins": [
+                {
+                        "type": "ptp",
+                        "ipMasq": true,
+                        "ipam": {
+                                "type": "host-local",
+                                "subnet": "172.16.30.0/24",
+                                "routes": [
+                                        {
+                                                "dst": "0.0.0.0/0"
+                                        }
+                                ]
+                        }
+                },
+                {
+                        "type": "portmap",
+						"capabilities": {"portMappings": true}
+                }
+        ]
+}
+```
+
+
+## Rule structure
+The plugin sets up two sequences of chains and rules - one "primary" DNAT
+sequence to rewrite the destination, and one additional SNAT sequence that
+rewrites the source address for packets from localhost. The sequence is somewhat
+complex to minimize the number of rules non-forwarded packets must traverse.
+
+
+### DNAT
+The DNAT rule rewrites the destination port and address of new connections.
+There is a top-level chain, `CNI-HOSTPORT-DNAT` which is always created and
+never deleted. Each plugin execution creates an additional chain for ease
+of cleanup. So, if a single container exists on IP 172.16.30.2 with ports 
+8080 and 8043 on the host forwarded to ports 80 and 443 in the container, the 
+rules look like this:
+
+`PREROUTING`, `OUTPUT` chains:
```

Comment:
Interesting. I'd always thought the `local` routing table was essentially read-only, but a quick experiment shows it's not (neat!).

Question: are you just concerned about performance, or is it expected that traffic from these interfaces are not forwarded? In a (non-k8s?) case, is it desirable that a GCE vip *would* be forwarded to a container? A quick glance at the documentation makes this at least seem plausible.

A few thoughts on possible solutions:
- Have a daemon that listens to netlink messages and updates a linux `ipset` with the list of local ifs
- Add a configuration option to the plugin to add conditions to the second-level entry rule. Then the administrator could exclude either specific CIDRs, devgroups, or ipsets
- Write the port reservation daemon, then match on `socket` (this is medium sketchy)

Either way, I'm thinking about adding a portset (via `ipset(8)`) to the top-level entry rule for improved performance.